lib/container/ContainerPlugin.js

```diff
@@ -0,0 +1,193 @@
+
+const { ConcatSource } = require("webpack-sources");
+const validateOptions = require("schema-utils");
+
+const PLUGIN_NAME = "ContainerPlugin";
+
+module.exports = class ContainerPlugin {
+	constructor(options) {
+		const name = options.name || "remoteEntry";
+
+		validateOptions(
+			{
+				type: "object",
+				properties: {
+					shared: {
+						type: "object"
+					},
+					expose: {
+						type: ["object", "array"]
+					},
+					name: {
+						type: "string",
+						default: name
+					},
+					library: {
+						type: "string",
+						default: name
+					},
+					libraryTarget: {
+						type: "string",
+						default: "global",
+						enum: [
+							"var",
+							"this",
+							"window",
+							"self",
+							"global",
+							"commonjs",
+							"commonjs2",
+							"amd",
+							"amd-require",
+							"umd",
+							"umd2",
+							"system"
+						]
+					},
+					filename: {
+						anyOf: [{ type: "string" }, { instanceof: "Function" }]
+					}
+				},
+				additionalProperties: false
+			},
+			options,
+			{ name: PLUGIN_NAME }
+		);
+
+		this.options = {
+			shared: options.shared || null,
+			name,
+			library: options.library || name,
+			libraryTarget: options.libraryTarget || "global",
+			filename: options.filename || undefined, // Undefined means, use the default behaviour
+			expose: options.expose || {}
+		};
+	}
+
+	apply(compiler) {
+		if (compiler.options.optimization.runtimeChunk) {
+			throw new Error(
+				"This plugin cannot integrate with RuntimeChunk plugin, please remote `optimization.runtimeChunk`."
+			);
+		}
+
+		compiler.options.output.jsonpFunction = `${
+			compiler.options.output.jsonpFunction
+		}${compiler.name || ""}${this.options.name}`;
+
+		compiler.hooks.make.tapAsync(PLUGIN_NAME, (compilation, callback) => {
+			let exposedMap = this.options.expose;
+
+			if (Array.isArray(this.options.expose)) {
+				exposedMap = {};
+				for (const exp of this.options.expose) {
+					// TODO: Check if this regex handles all cases
+					exposedMap[exp.replace(/(^(?:[^\w])+)/, "")] = exp;
+				}
+			}
+
+			compilation.addEntry(
+				compilation.context,
+				new ContainerEntryDependency(
+					Object.entries(exposedMap).map(([name, request], idx) => {
+						const dep = new ContainerExposedDependency(name, request);
+						dep.loc = {
+							name,
+							index: idx
+						};
+						return dep;
+					}),
+					this.options.name
+				),
+				this.options.name,
+				callback
+			);
+		});
+
+		compiler.hooks.thisCompilation.tap(
+			PLUGIN_NAME,
+			(compilation, { normalModuleFactory }) => {
+				compilation.dependencyFactories.set(
+					ContainerEntryDependency,
+					new ContainerEntryModuleFactory()
+				);
+
+				compilation.dependencyFactories.set(
+					ContainerExposedDependency,
+					normalModuleFactory
+				);
+
+				const renderHooks = JavascriptModulesPlugin.getCompilationHooks(
+					compilation
+				);
+
+				renderHooks.render.tap(PLUGIN_NAME, (source, { chunk }) => {
```

Comment:
This is fine for now, later we can migrate to the new library system once this is done.